pkg/agent/util/ndp/ndp_linux.go

```diff
@@ -0,0 +1,161 @@
+	// Target Link-Layer Address                    2
+	// Prefix Information                           3
+	// Redirected Header                            4
+	// MTU                                          5
+	ndpOptionType = 2
+
+	// Minimum byte length values for each type of valid Message.
+	naLen = 20
+
+	// expected IPv6 hop limit, hop limit is always 255 for all NDP messages, RFC 4861.
+	hopLimit = 255
+)
+
+func checkIPv6(ip net.IP) error {
+	if ip.To16() == nil || ip.To4() != nil {
+		return fmt.Errorf("invalid IPv6 address: %s", ip.String())
+	}
+	return nil
+}
+
+// NeighborAdvertisement sends an NDP Neighbor Advertisement over interface 'iface' from 'srcIP'.
+func NeighborAdvertisement(srcIP net.IP, iface *net.Interface) error {
+	if err := checkIPv6(srcIP); err != nil {
+		return err
+	}
+
+	addrs, err := iface.Addrs()
+	if err != nil {
+		return fmt.Errorf("interface address error: %v", err)
+	}
+
+	ipAddr := &net.IPAddr{}
+
+	for _, a := range addrs {
+		ipn, ok := a.(*net.IPNet)
+		if !ok {
+			continue
+		}
+
+		if err := checkIPv6(ipn.IP); err != nil {
+			continue
+		}
+
+		if ipn.IP.IsLinkLocalUnicast() {
+			ipAddr = &net.IPAddr{
+				IP:   ipn.IP,
+				Zone: iface.Name,
+			}
+			break
+		}
+	}
+
+	ic, err := icmp.ListenPacket("ip6:ipv6-icmp", ipAddr.String())
+	if err != nil {
+		return fmt.Errorf("listen icmp error: %v", err)
+	}
+
+	pc := ic.IPv6PacketConn()
+	if err := pc.SetHopLimit(hopLimit); err != nil {
+		return fmt.Errorf("ipv6 conn set hop limit error: %v", err)
+	}
+	if err := pc.SetMulticastHopLimit(hopLimit); err != nil {
+		return fmt.Errorf("ipv6 conn set multicast hoplimit error: %v", err)
+	}
+
+	// Calculate and place ICMPv6 checksum at correct offset in all messages.
+	const chkOff = 2
+	if err := pc.SetChecksum(true, chkOff); err != nil {
+		return fmt.Errorf("ipv6 conn set checksum error: %v", err)
+	}
+
+	defer pc.Close()
+	mb, err := newNDPNeighborAdvertisementMessage(srcIP, iface.HardwareAddr)
+	if err != nil {
+		return fmt.Errorf("new NDP Neighbor Advertisement Message error: %v", err)
+	}
+
+	cm := &ipv6.ControlMessage{
+		HopLimit: hopLimit,
+		Src:      ipAddr.IP,
+		IfIndex:  iface.Index,
+	}
+	dstAddr := &net.IPAddr{
+		IP:   net.IPv6linklocalallnodes,
+		Zone: iface.Name,
+	}
+	if _, err = pc.WriteTo(mb, cm, dstAddr); err != nil {
+		return fmt.Errorf("writing a payload of the IPv6 datagram error: %v", err)
+	}
+	return nil
+}
+
+func newNDPNeighborAdvertisementMessage(targetAddress net.IP, hwa net.HardwareAddr) ([]byte, error) {
+	naMsgBytes := make([]byte, naLen)
+	naMsgBytes[0] |= 1 << 5
+	copy(naMsgBytes[4:], targetAddress)
+
+	marshall := func() ([]byte, error) {
+		if 1+1+len(hwa) != int(ndpOptionLen*8) {
+			return nil, fmt.Errorf("hardwareAddr length error: %s", hwa)
+		}
+		b := make([]byte, ndpOptionLen*8)
+		b[0] = ndpOptionType
+		b[1] = ndpOptionLen
+		copy(b[2:], hwa)
+		return b, nil
+	}
+
+	optionsBytes, err := marshall()
+	if err != nil {
+		return nil, err
+	}
```

Comment:
It seems the `marshall` function is pointless, which adds 6 extra lines of code but neither improve the readability much (the whole function is about creating a message, why the options bytes are extracted specially with a generic name "marshall") nor be shared.

Reply:
done